packages/medusa/src/services/discount-condition.ts

```diff
@@ -0,0 +1,238 @@
+    this.discountRepository_ = discountRepository
+
+    /** @private @const {DiscountConditionRepository} */
+    this.discountConditionRepository_ = discountConditionRepository
+
+    /** @private @const {EventBus} */
+    this.eventBus_ = eventBusService
+  }
+
+  withTransaction(transactionManager: EntityManager): DiscountConditionService {
+    if (!transactionManager) {
+      return this
+    }
+
+    const cloned = new DiscountConditionService({
+      manager: transactionManager,
+      discountRuleRepository: this.discountRuleRepository_,
+      discountRepository: this.discountRepository_,
+      discountConditionRepository: this.discountConditionRepository_,
+      eventBusService: this.eventBus_,
+    })
+
+    cloned.transactionManager_ = transactionManager
+
+    return cloned
+  }
+
+  async retrieve(
+    conditionId: string,
+    discountId: string
+  ): Promise<DiscountCondition & { discount: Discount }> {
+    return await this.atomicPhase_(async (manager: EntityManager) => {
+      const conditionRepo = manager.getCustomRepository(
+        this.discountConditionRepository_
+      )
+
+      // left join with discount to ensure we are working with correct comb. of condition and discount
+      const condition = await conditionRepo.findOneWithDiscount(
+        conditionId,
+        discountId
+      )
+
+      // slightly different error message than usual accouting for the left join
+      if (!condition || !condition?.discount) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `DiscountCondition with id ${conditionId} was not found for Discount ${discountId}`
+        )
+      }
+
+      return condition
+    })
+  }
+
+  static resolveConditionType_(data: UpsertDiscountConditionInput):
+    | {
+        type: DiscountConditionType
+        resource_ids: string[]
+      }
+    | undefined {
+    switch (true) {
+      case !!data.products?.length:
+        return {
+          type: DiscountConditionType.PRODUCTS,
+          resource_ids: data.products!,
+        }
+      case !!data.product_collections?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_COLLECTIONS,
+          resource_ids: data.product_collections!,
+        }
+      case !!data.product_types?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TYPES,
+          resource_ids: data.product_types!,
+        }
+      case !!data.product_tags?.length:
+        return {
+          type: DiscountConditionType.PRODUCT_TAGS,
+          resource_ids: data.product_tags!,
+        }
+      case !!data.customer_groups?.length:
+        return {
+          type: DiscountConditionType.CUSTOMER_GROUPS,
+          resource_ids: data.customer_groups!,
+        }
+      default:
+        return undefined
+    }
+  }
+
+  async upsertCondition(
+    discountIdOrDiscount: string | Discount,
+    data: UpsertDiscountConditionInput
+  ): Promise<void> {
+    let resolvedConditionType
```

Comment:
This seams to be a const that can be declare line 138 if I did not miss something :)

Reply:
No, it's used in the error handler of the atomic phase, so needs to be defined outside of the work scope